src/nimble_contact_manager.cc

```diff
@@ -618,971 +394,1343 @@ namespace nimble {
+  auto num_contacts = numActiveContactFaces();
+  global_data.push_back(static_cast<double>(num_contacts));
+
+  std::vector<int> const &block_ids = mesh.GetBlockIds();
+  for (auto &block_id : block_ids) {
+    elem_data_labels_for_output[block_id] = std::vector<std::string>();
+    derived_elem_data_labels[block_id] = std::vector<std::string>();
+  }
+
+  // node_data_for_output contains displacement_x, displacement_y,
+  // displacement_z
+  auto num_nodes = mesh.GetNumNodes();
+  for (auto &ndata : node_data_for_output)
+    ndata.resize(num_nodes);
+  const double *model_coord_x = mesh.GetCoordinatesX();
+  const double *model_coord_y = mesh.GetCoordinatesY();
+  const double *model_coord_z = mesh.GetCoordinatesZ();
+
+  size_t nfaces = numContactFaces();
+
+  int node_index(0);
+  for (size_t i_face = 0; i_face < nfaces; i_face++) {
+    ContactEntity const &face = getContactFace(i_face);
+    auto contact_status = static_cast<double>(face.contact_status());
+    node_data_for_output[0][node_index] =
+        face.coord_1_x_ - model_coord_x[node_index];
+    node_data_for_output[1][node_index] =
+        face.coord_1_y_ - model_coord_y[node_index];
+    node_data_for_output[2][node_index] =
+        face.coord_1_z_ - model_coord_z[node_index];
+    node_data_for_output[3][node_index] = contact_status;
+    node_index += 1;
+    node_data_for_output[0][node_index] =
+        face.coord_2_x_ - model_coord_x[node_index];
+    node_data_for_output[1][node_index] =
+        face.coord_2_y_ - model_coord_y[node_index];
+    node_data_for_output[2][node_index] =
+        face.coord_2_z_ - model_coord_z[node_index];
+    node_data_for_output[3][node_index] = contact_status;
+    node_index += 1;
+    node_data_for_output[0][node_index] =
+        face.coord_3_x_ - model_coord_x[node_index];
+    node_data_for_output[1][node_index] =
+        face.coord_3_y_ - model_coord_y[node_index];
+    node_data_for_output[2][node_index] =
+        face.coord_3_z_ - model_coord_z[node_index];
+    node_data_for_output[3][node_index] = contact_status;
+    node_index += 1;
+  }
+
+  const size_t nnodes = numContactNodes();
+  for (size_t i_node = 0; i_node < nnodes; i_node++) {
+    ContactEntity const &node = getContactNode(i_node);
+    node_data_for_output[0][node_index] =
+        node.coord_1_x_ - model_coord_x[node_index];
+    node_data_for_output[1][node_index] =
+        node.coord_1_y_ - model_coord_y[node_index];
+    node_data_for_output[2][node_index] =
+        node.coord_1_z_ - model_coord_z[node_index];
+    node_data_for_output[3][node_index] =
+        static_cast<double>(node.contact_status());
+    node_index += 1;
+  }
+
+  out.WriteStep(t, global_data, node_data_for_output,
+                elem_data_labels_for_output, elem_data_for_output,
+                derived_elem_data_labels, derived_elem_data);
+}
+
+
+//
+// Protected functions
+//
+
+void ContactManager::ApplyDisplacements(const double *displacement) {
```

Comment:
thanks, I really happy to see that this object is now composed of shorter functions that (more or less) do not scroll off the screen

Reply:
@rjones-sandia 
Great!
I tried to keep the modifications for `nimble_contact_manager.cc` as limited as possible.
If I remembered correctly, we will update the `ContactManager` hierarchy.
So I focused on making everything work.